compiler/rustc_typeck/src/check/method/suggest.rs

```diff
@@ -1129,6 +1129,37 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     label_span_not_found();
                 }
 
+                if let SelfSource::MethodCall(expr) = source
+                    && let ty::Adt(kind, substs) = actual.kind()
+                    && tcx.get_diagnostic_item(sym::Result) == Some(kind.did())
+                {
+                    let call_expr =
+                        self.tcx.hir().expect_expr(self.tcx.hir().get_parent_node(expr.hir_id));
+                    let ok_ty = substs.type_at(0);
+                    if let Ok(pick) =
+                        self.lookup_probe(span, item_name, ok_ty, call_expr, ProbeScope::AllTraits)
+                    {
+                        err.span_note(
+                            tcx.def_span(pick.item.def_id),
+                            &format!("the method `{item_name}` exists on the type `{ok_ty}`"),
+                        );
+                        if let Some(ret_ty) = self
+                            .ret_coercion
+                            .as_ref()
+                            .map(|c| self.resolve_vars_if_possible(c.borrow().expected_ty()))
+                            && let ty::Adt(kind, _) = ret_ty.kind()
+                            && tcx.get_diagnostic_item(sym::Result) == Some(kind.did())
```

Comment:
This is checking that the return type is something that `?` can coerce to, right? If the return type isn't that, maybe we can 1) suggest changing the return type and 2) use `.unwrap()`. If we do 2), then we should also support this for `Option`. We can also build some scaffolding to suggest using `if let`, which could work for *any* ADTs...

Reply:
Hmmm... I'm not sure if I want to encourage people to .unwrap, though I don't know if we have precedent for that other suggestions. Or at least, maybe I'll suggest expect.

Reply:
If we can do the more involved `if let` suggestion, that'd be fine by me to skip the `unwrap`.

Reply:
> If we can do the more involved `if let` suggestion

Are you suggesting we transform something like:

```rust
result.call();
```

into

```rust
if let Ok(value) = result {
  value.call();
}
```

Because if so, I'd probably want to reserve that for a follow-up PR. I need to think of heuristics for when it's meaningful to do so, because I wouldn't want to suggest it for any nested expression, but only ones that are reasonably statement-like on their own. Like not suggesting it when the bad method call is here:

```rust
let closure = || call_something(result_needs_unwrap.call(), other_expression());
//                              ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Reply:
Yes, and point taken. As follow up work what we could so is check if the current ty is an enum, and prove each variant that has a single value for the type and emit a note saying "the variant `Result::Ok(Ty)` has method `foo`" (but still special case for `Result` and `Option`).

Reply:
Yeah, I've been thinking of this a bit more, and I would also like to generalize some sort of message to more enums somehow in a follow-up PR.

Let me special-case it for `Result` and `Option` in this PR, along with suggesting `.expect` (and/or changing the return type of the function) when we don't return `Result`.